File: bus_network.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAXSTOPS 310
#define MAXNODES 6000
#define MAXCOST 30000

int node_ref[MAXSTOPS][MAXNODES];
int g[MAXNODES][MAXNODES];
int visited[MAXNODES];
int nodeq[MAXNODES];
int costs[MAXNODES];    

int main(int argc, char ** argv)
{
    char line[1000];
    FILE * f = fopen(argv[1], "r");
    if(f == NULL) {
        exit(0);
    }
    int src, dst, i, j, k, read_cnt, offset, rti, nodei, start, station;
    int cnt, tmp, ans, res;
    int routes[500][200];
    char * rt, * c, *node, *data; 
    while(fgets(line, sizeof(line), f) != NULL) {
        line[strlen(line)-1] = '\0';
        rt = strtok(line, "; ");
        sscanf(rt, "(%d,%d)", &src, &dst);
        //printf("src: %d, dst: %d\n", src, dst);
        i = 0;
        while(rt = strtok(NULL, "; ")) {
            //printf("ugly route: %s\n", rt);
            j = 1;
            node = strchr(rt, ']');
            *node = '\0';
            node = strchr(rt, '[');
            node++;
            offset = 0;
            //printf("node %s\n", node);
            offset = 0;
            tmp = strlen(node);
            while(offset < tmp && 1==sscanf(node+offset, "%d,%n", &routes[i][j++], &read_cnt)) {
                offset += read_cnt;
            }
            /*printf("route %d\n", i);
            for(k=0;k<j;k++) {
                printf("%d ", routes[i][k]);
            }
            printf("\n");*/
            routes[i][0] = j-1;
            i++;
        }
        memset(node_ref, 0, MAXSTOPS * MAXNODES);
        for(k=0;k<MAXNODES;k++) {
            for(j=0;j<MAXNODES;j++) {
                g[k][j] = MAXCOST;
            }
        }
        start = 0;
        for(rti=0;rti<i;rti++) {
            //printf("route %d, count %d\n", rti, routes[rti][0]);
            for(nodei=1;nodei<=routes[rti][0];nodei++) {
                if(nodei < routes[rti][0]) {
                    g[start+nodei][start+nodei+1] = 7;
                    g[start+nodei+1][start+nodei] = 7;
                }
                //printf("nodei %d\n", nodei);
                station = routes[rti][nodei];
                cnt = node_ref[station][0];
                //printf("station %d, cnt %d\n", station, cnt);
                for(k=1;k<=cnt;k++) {
                    tmp = node_ref[station][k];
                    g[start+nodei][tmp] = 12;
                    g[tmp][start+nodei] = 12;
                }
                node_ref[station][cnt+1] = start+nodei;
                node_ref[station][0] = cnt+1;
            }   
            start += routes[rti][0];    
        }
        /*printf("create graph\n");
        for(i=0;i<MAXNODES;i++) {
            for(j=0;j<MAXNODES;j++) {
                if(g[i][j] < MAXCOST) {
                    printf("(%d %d):%d ", i, j, g[i][j]);
                }
            }
        }
        printf("\n");*/
        ans = MAXCOST;
        for(i=1;i<=node_ref[src][0];i++) {
            res = get_cost(node_ref[src][i], node_ref[dst]);
            if(res<ans) ans = res;
        }
        if(ans < MAXCOST) {
            printf("%d\n", ans);
        } else {
            printf("None\n");
        }
    }
    fclose(f);
    return 0;
}

int get_cost(int src, int dst[]) {
    int end = 0, node, head = 0, i, adj, d;
    int ans = MAXCOST;
    nodeq[end++] = src;
    for(i=0;i<MAXNODES;i++) {
        costs[i] = MAXCOST;
        visited[i] = 0;
    }
    costs[src] = 0;
    while(head<end) {
        node = nodeq[head++];
        visited[node] = 1;
        for(i=1;i<MAXNODES;i++) {
            if(g[node][i] < MAXCOST && visited[i] == 0) {
                nodeq[end++] = i;
                if (costs[i] > costs[node]+g[node][i]) {
                    costs[i] = costs[node]+g[node][i];
                }
            }
        }
    }
    for(i=1;i<=dst[0];i++) {
        d = dst[i];
        if(ans>costs[d])ans = costs[d];
    }
    return ans;    
}

File: bus_network.py
import sys
import profile
import array
import collections

def main():
    with open(sys.argv[1], "r") as f:
        for line in f.readlines():
            data = line.strip().split('; ') 
            src, dst = map(int, data[0].strip('()').split(','))
            routes = [map(int, r.split('=')[1].strip('[]').split(',')) for r in data[1:]]
            #create a graph with the routes 
            #print "routes: %s" % routes
            start = 0
            g = {}
            node_ref = {}
            for r in routes:
                rlen = len(r)
                for i in range(rlen):
                    if i < rlen-1:
                        if start+i not in g:
                            g[start+i] = [(start+i+1, 7)]
                        else:
                            g[start+i].append((start+i+1, 7))
                        if start+i+1 not in g:
                            g[start+i+1] = [(start+i, 7)]
                        else:
                            g[start+i+1].append((start+i, 7))
                    if r[i] in node_ref:
                        for node in node_ref[r[i]]:
                            g[start+i].append((node, 12))
                            g[node].append((start+i, 12))
                    if r[i] not in node_ref:
                        node_ref[r[i]] = [start+i]
                    elif start+i not in node_ref[r[i]]:
                        node_ref[r[i]].append(start+i)
                start += rlen
            #print "create graph: %s" % g
            ans = 100000000
            #print node_ref[src]
            for s in node_ref[src]:
                visited = [False] * start
                costs = [100000000] * start
                costs[s] = 0
                get_cost(s, g, visited, costs)
                res = min([costs[node] for node in node_ref[dst]])
                if res < ans:ans = res
            if ans == 100000000:
                print "None"
            else:
                print ans
    sys.exit(0)

def get_cost(src, g, visited, costs):
    nq = collections.deque()
    nq.append(src)
    costs[src] = 0
    while nq:
        curnode = nq.popleft()
        visited[curnode] = True
        for (node, w) in g[curnode]:
            if not visited[node]:
                nq.append(node)
                if costs[node] > costs[curnode] + w:
                    costs[node] = costs[curnode] + w 

def test():
    g = {0: [(1, 7)], 1: [(0, 7), (2, 7)], 2: [(1, 7)], 3: [(4, 7), (11, 12)], 4: [(3, 7), (5, 7)], 5: [(4, 7), (6, 7)], 6: [(5, 7), (7, 7)], 7: [(6, 7)], 8: [(9, 7)], 9: [(8, 7), (10, 7)], 10: [(9, 7), (11, 7)], 11: [(10, 7), (3, 12)]}
    visited = [False] * 12
    costs = [100000000] * 12
    s = 3
    get_cost(s, g, visited, costs)
    print "src %d, costs: %s" % (s, costs) 
    #get_cost(3, g, visited, costs)
    #print "src 3, costs: %s" % 

if __name__ == '__main__':
    #profile.run("main();")
    main()
